File: csharp/src/Eunomia.Server.Api/Models/PacketEnvelope.cs
// Copyright (c) 2026, zannagh. All rights reserved.
// See License in the project root for license information.

using System.Text.Json;

namespace Eunomia.Server.Api.Models;

/// <summary>
/// Wire envelope shared with the Java client, used for both REST puts and websocket pushes.
/// <see cref="Payload"/> is opaque to the server: it is stored/relayed verbatim.
/// </summary>
public sealed record PacketEnvelope
{
    /// <summary>MC-server identity, e.g. "mc.hypixel.net:25565"; partitions all data.</summary>
    public required string Scope { get; init; }

    /// <summary>"namespace:path" channel identifier.</summary>
    public required string Channel { get; init; }

    /// <summary>Slash-joined KeyPath string for keyed packets; null for plain packets.</summary>
    public string? Key { get; init; }

    /// <summary>True: store + push-on-connect + relay. False: relay only.</summary>
    public required bool Replicated { get; init; }

    /// <summary>Player UUID (canonical MC form, dashed lowercase).</summary>
    public required string Sender { get; init; }

    /// <summary>The DTO JSON, opaque to the server.</summary>
    public required JsonElement Payload { get; init; }
}

File: java/core/src/main/java/de/zannagh/eunomia/clients/PacketEnvelope.java
package de.zannagh.eunomia.clients;

import com.google.gson.JsonElement;

/**
 * The JSON envelope exchanged with the external relay (the C# server), over both the REST send path and the
 * WebSocket receive path. Field names and shape match the server's {@code PacketEnvelope} exactly.
 * <p>
 * {@link #scope} partitions all data by the Minecraft server the client is connected to, so two clients on
 * different servers never see each other's data. {@link #name} carries the human-readable server label (the
 * player's entry for that server) purely for display on the server side; it never affects partitioning.
 * {@link #key} is the {@code KeyPath} (slash-joined) for keyed packets and {@code null} for plain ones;
 * {@link #replicated} marks a {@code Replicated} payload the server must store and push to newcomers.
 */
public class PacketEnvelope {

    public String scope;

    public String name;

    public String channel;

    public String key;

    public boolean replicated;

    public String sender;

    public JsonElement payload;

    public PacketEnvelope() {
    }

    public PacketEnvelope(String scope, String name, String channel, String key, boolean replicated, String sender, JsonElement payload) {
        this.scope = scope;
        this.name = name;
        this.channel = channel;
        this.key = key;
        this.replicated = replicated;
        this.sender = sender;
        this.payload = payload;
    }
}
